### DISCUSSION TITLE
Transmitting of the tooltype

### DISCUSSION THREAD

**Discussion by rhmls** [umati/UA-for-MachineTools#28] [Team Posts]
I need help on how to provide the tooltype correctly via umati.
Here is some background information:

In our control we have an internal unique ID for every tool. 
In addition, each tool has another ID which is can be defined by the customer (barcode/datamatrixcode) and a name, which is also unique and can be defined individualy.
Tools can be related to a tooltype, if they are identical and shall share typical milling parameters.

Every tooltype has a unique ID which is only for internal use and is not published provided to the outside.
Each tooltype also has a unique name, which is used to automatically select and get a proper tool from the magazine.

For publishing this data via umati, we are currently using the NumberBased ControlIdentifierInterpretation, because providing the internal unique ID of the tooltype would be of no use for the customer.
So the variable „ControlIdentifier1“ is used for the unique ID of the tool and „Identifier“ is used for the ID defined by the customer.

As I see it, the only option to transmit the tooltype is via the variable „Name“, where I’d build a string containing both, the name of the tool and the name of the tooltype, so the tool can be easily recognized.
I don’t think this is the best option, so I’d be very thankful for any advice on that issue!

From our point of view, it might be a better solution, if we could publish the name of the related tooltype in a separate node as a string.

@GoetzGoerisch told me that you (@TonjaHeinemann and @JensDamm) might know something about this issue.

Thanks in advance!

**Comment by GoetzGoerisch** [comment]
Dear @rhmls,

thank you for this question.

The spec says in https://reference.opcfoundation.org/MachineTool/docs/8.5.4/:
_In a system with a ToolGroupBasedManagemet, tools are identified by a group and an indexing number inside this group, which are provided as ControlIdentifier1 and ControlIdentifier2 respectively._

Isn't this your use case?

Furthermore it says:
_Should yet another approach be present in a given system, this is indicated by the ControlIdentifierInterpretation being reported as custom. It shall be noted that this identification data is used to identify the tool in the reference frame of the tool management system inside the machine. In many applications the machine control system uses these identifiers to handle multiple tools which are equivalent and present in the machine as spares. For an identification of the tool inside the NC program or globally, independent of the machine management view, the properties Name and Identifier are provided by the BaseToolType._

If I misunderstood, could you give a concrete example to visualize the approach?

@JensDamm please provide your view?

**Comment by JensDamm** [comment]
@rhmls 
My suggestion is to
- Write the tool type/group to ControlIdentifier2
- Set the interpretation either to numberBased or to custom.
@GoetzGoerisch 
The intention for group based is to use both identifiers to identify a tool. So I would like to avoid to use it here.

The general idea with the _ControlIdentifierInterpretation_ is to tell the client how to uniquely identify a tool on the machine tool.
So even if _ControlIdentifier2_ is given the provided information are still valid using "numberBased".
If you set it to custom you have to define how the client must interpret the information, would be 100% formal correct.
I personally would prefer to set it to number based, since it's still posiible to identify the tool on the machine with 
ControlIdentifier1.

Since you mention that you would like to provide the "toolType" as a string, could you tell which kind of control you are using?

To understand your wording let's say e.g. a ball mill is a tool type. Drills would be a tool class.
From my point I would reference a  group of tools with similar properties as group.

**Comment by rhmls** [comment]
Hi Jens, thank you for your response. Somehow this issue got lost in my tasks...
We are using a control called "RMS6" which is developed by ourselves. 

Here is an example:
A Tool has a unique ID and a tool type. A tool type can be anything that is defined by the user. A valid tool type would be for example "D0k2_r0k1". You could also name a tool type "MyToolType123". 
But because ControlIdentifier2 is a UInt32 I can not just change it to a string, can I?

**Comment by GoetzGoerisch** [comment]
_ControlIdentifier2_ is _UInt32_, thus cannot be changed to a _String_.
@rhmls I think the approach described was not thought of in the first version. @IOTeule you look into this for version 1 right?
